MyOnboarding/src/MyOnboardingApp.Services/Services/UpdateItemService.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System;
+using System.Threading.Tasks;
+using MyOnboardingApp.Contracts.Generators;
+using MyOnboardingApp.Contracts.Models;
+using MyOnboardingApp.Contracts.Repository;
+using MyOnboardingApp.Contracts.Services;
+using MyOnboardingApp.Contracts.Validation;
+
+namespace MyOnboardingApp.Services.Services
+{
+    internal class UpdateItemService : AbstractUpsertService, IUpdateItemService
+    {
+        private readonly ITodoListRepository _repository;
+        private readonly IDateTimeGenerator _dateTimeGenerator;
+
+
+        public UpdateItemService(ITodoListRepository repository, IDateTimeGenerator dateTimeGenerator, IInvariantValidator<TodoListItem> validator)
+            : base(validator)
+        {
+            _dateTimeGenerator = dateTimeGenerator;
+            _repository = repository;
+        }
+
+
+        public async Task<IItemWithErrors<TodoListItem>> EditItemAsync(TodoListItem replacingItem, IResolvedItem<TodoListItem> existingItem)
+        {
+            if (replacingItem == null)
+            {
+                throw new ArgumentNullException(nameof(replacingItem), "Item to complete must not be null.");
+            }
+
+            if (existingItem == null || !existingItem.WasOperationSuccessful)
+            {
+                throw new ArgumentException("Existing item does not actually exist.", nameof(existingItem));
+            }
+
+            var initData = GetGeneratedData(existingItem.Item);
+
+            return await TryCompleteAndStoreItemAsync(replacingItem, initData);
+        }
+
+
+        protected override async Task StoreToDatabase(TodoListItem completedItem)
+            => await _repository.ReplaceItemAsync(completedItem);
+
+
+        protected virtual ItemInitializingData GetGeneratedData(TodoListItem item)
```

Comment:
Why is this `protected virtual`?

Reply:
Speaking about it, is this class supposed to be inherited, like ever? 😛 

I don't want you to `seal` all classes that are meant as "leaf" implementations, especially since they are all `internal`. But, please, bear in mind that you should always have any of your public classes either `sealed` or enabled for meaningful inheritance (i.e. have at least one `virtual` or `abstract` method that somebody might want to `override`)…

Reply:
Why doesn't this class accept `IResolvedItem<TodoListItem>`? (https://github.com/petrareb/kentico-onboarding-cs/pull/4#discussion_r263394536 – the "all the way down" part)